### ISSUE TITLE
swagger-models - ObjectProperty children are lost - no way to parse nested object definitions that are not $refs

### ISSUE THREAD

**Issue by benken-parasoft** [swagger-api/swagger-core#1202]
I am using swagger-models to marshall Swagger back and forth.

When consuming a Swagger having a JSON schema object definition where one of the properties is another object definition, all properties in the nested object are lost.  The nested object definition becomes an ObjectProperty but there is no method to get the children.  Marshalling the ModelImpl back to JSON (round trip) shows the properties of the nested object being lost.

There appears to be no way to use swagger-models to parse the nested object definition!  This causes a problem for any application that wishes to accurately consume Swagger documents and their embedded JSON Schemas.

Below is a JUnit that demonstrates the problem with ObjectProperty:

<pre><code>import static org.junit.Assert.*;

import com.fasterxml.jackson.core.*;
import com.fasterxml.jackson.databind.*;

import io.swagger.models.*;
import io.swagger.models.properties.*;
import io.swagger.util.*;

import java.io.*;
import java.util.*;

import org.junit.*;

public class SwaggerTest {
    @Test
    public void testMinMaxItems() throws JsonProcessingException, IOException {
        ObjectMapper mapper = Yaml.mapper();
        JsonNode rootNode = mapper.readTree(SAMPLE_SWAGGER);
        assertNotNull(rootNode);
        Swagger doc = mapper.convertValue(rootNode, Swagger.class);
        Map<String, Model> definitions = doc.getDefinitions();
        assertEquals(3, definitions.size());

        // What happened to children of submodel?!?
        Model model = definitions.get("Model");
        assertEquals(ModelImpl.class, model.getClass());
        ModelImpl modelImpl = (ModelImpl) model;
        Map<String, Property> properties = modelImpl.getProperties();
        assertEquals(2, properties.size());
        Property property = properties.get("simple1");
        assertEquals(IntegerProperty.class, property.getClass());
        property = properties.get("submodel");
        assertEquals(ObjectProperty.class, property.getClass());
        //ObjectProperty objectProperty = (ObjectProperty) property;
        //objectProperty.getProperties()
        assertEquals("{\"type\":\"object\",\"required\":[\"simple1\",\"submodel\"],\"properties\":{\"simple1\":{\"type\":\"integer\",\"format\":\"int32\"},\"submodel\":{\"type\":\"object\"}}}",
                mapper.valueToTree(model).toString());
        System.out.println(SAMPLE_SWAGGER);
    }

    private static final String SAMPLE_SWAGGER =
            "swagger: '2.0'\n" +
            "info:\n" +
            "  version: '1.0.0'\n" +
            "  title: Simple test\n" +
            "host: fakehostname\n" +
            "basePath: /fakepath\n" +
            "paths:\n" +
            "  /Test:\n" +
            "    post:\n" +
            "      parameters:\n" +
            "        - in: body\n" +
            "          name: body\n" +
            "          schema:\n" +
            "            $ref: \"#/definitions/Model\"\n" +
            "      responses:\n" +
            "        200: \n" +
            "          description: OK  \n" +
            "definitions:\n" +
            "  Model:\n" +
            "    type: object\n" +
            "    required:\n" +
            "      - simple1\n" +
            "      - submodel\n" +
            "    properties:\n" +
            "      simple1:\n" +
            "        type: integer\n" +
            "      submodel:\n" +
            "        required:\n" +
            "          - element1\n" +
            "        type: object\n" +
            "        properties:\n" +
            "          element1:\n" +
            "            type: integer\n" +
            "            format: int32\n" +
            "  Model2:\n" +
            "    type: object\n" +
            "    required:\n" +
            "      - simple1\n" +
            "      - submodel\n" +
            "    properties:\n" +
            "      simple1:\n" +
            "        type: integer\n" +
            "      submodel:\n" +
            "        $ref: \"#/definitions/SubModel\"\n" +
            "  SubModel:\n" +
            "    type: object\n" +
            "    properties:\n" +
            "      element1:\n" +
            "        type: integer\n" +
            "        format: int32\n" +
            "";
}</code></pre>

**Comment by jasonh-n-austin** [comment]
+1 to this, ran across it today. We dereference our schema (for other broken libraries), rather than using `$ref`. As such, any child objects in the schema just disappear in `swagger-core`.

**Comment by benken-parasoft** [comment]
On a side note, other consumers don't seem to have a problem.  For example, the online Swagger-editor (http://editor.swagger.io/) is able to display the models.  The swagger-core/models java API is just unable to consume the nested object definition.

**Comment by jasonh-n-austin** [comment]
@benken-parasoft AFAIK that's all powered by `swagger-js` (Javascript/Node-based).

**Comment by bchazalet** [comment]
same issue here. But I wonder then, how do people parse their swagger in scala/java then? Is it that most people don't have the need (or do it in another language)? Or do they rely on the json directly?

Doesn't that make the whole thing useless if you can't properly read into an ObjectProperty?

**Comment by fehguy** [comment]
@bchazalet yes, this is a missing feature.  The spec allows it and it will be implemented.  It's just not complete.

In the mean time, you can flatten you model references.  For example your `Model` has a nested schema in a property--instead of nesting, you can use a ref:

```yaml
Model:
  type: object
  required:
    - simple1
    - submodel
  properties:
    simple1:
      type: integer
    submodel:
      required:
        - element1
      type: object
      properties:
        element1:
          type: integer
          format: int32
```

can be this:

```yaml
Model:
  type: object
  required:
    - simple1
    - submodel
  properties:
    simple1:
      type: integer
    submodel:
      $ref: '#/definitions/SubModel'
SubModel:
  required:
    - element1
  type: object
  properties:
    element1:
      type: integer
      format: int32
```

**Comment by benken-parasoft** [comment]
Also see the example from my JUnit in comment.  "Model1" has a nested object definition and can't be parsed completely.  The nested object definition is lost.  In contrast, "Model2" uses a "$ref" for the nested object definition instead.

Unfortunately, I can't necessarily tell my users they have to change their document just because my code can't parse it.  Perhaps a workaround could be to pre-process the document somehow to "normalize" it before unmarshalling it with swagger-models.  Not sure what would be involved with doing that.  Not sure what would be easier, writing a workaround like that or addressing the problem directly.  :)

**Comment by fehguy** [comment]
Fixing the underlying issue is not terribly difficult.

What would you expect the embedded object to be "named"?  For example, a code generation will likely create a second object for the complex property.  Since we probably wouldn't want it to be an inner class, what would you expect it to be generated as, as an end user?

**Comment by bchazalet** [comment]
@fehguy yes naming is probably the most annoying part, but I would argue that this is not something we should care about in core. Core should be as close to the specs as possible and if codegen wants to ignore some parts, that's fair enough but that should happen in the codegen project.

I am saying all this because in my use case I want to parse the swagger definitions to build internal tools on top of it (think of API gateway such as [tyk](https://tyk.io/) or [Kong](https://github.com/mashape/kong)) so I need to be able to parse any valid definitions not just the ones that are 'generable'.

**Comment by fehguy** [comment]
I get it.  Yes, let's get that supported in core, I'll bump up the priority on this.

**Comment by bchazalet** [comment]
@fehguy that's great. I have submitted PR #1314 for it. Let me know if you think it can be improved, I'll be happy to take it even further.

**Comment by frantuma** [comment]
Closing, fixed in #1314.
